File: src/ts/magma/app/ValueCompiler.ts
/*[
	JVMList: jvm.api.collect.list, 
	Lists: jvm.api.collect.list, 
	Files: jvm.api.io, 
	Actual: magma.annotate, 
	Namespace: magma.annotate, 
	Collector: magma.api.collect, 
	EmptyHead: magma.api.collect.head, 
	FlatMapHead: magma.api.collect.head, 
	Head: magma.api.collect.head, 
	HeadedIter: magma.api.collect.head, 
	MapHead: magma.api.collect.head, 
	RangeHead: magma.api.collect.head, 
	SingleHead: magma.api.collect.head, 
	Iter: magma.api.collect, 
	Iters: magma.api.collect, 
	Joiner: magma.api.collect, 
	Iterable: magma.api.collect.list, 
	List: magma.api.collect.list, 
	ListCollector: magma.api.collect.list, 
	Console: magma.api.io, 
	IOError: magma.api.io, 
	Path: magma.api.io, 
	None: magma.api.option, 
	Option: magma.api.option, 
	Some: magma.api.option, 
	Err: magma.api.result, 
	Ok: magma.api.result, 
	Result: magma.api.result, 
	Characters: magma.api.text, 
	Strings: magma.api.text, 
	Tuple2: magma.api, 
	Tuple2Impl: magma.api, 
	Application: magma.app, 
	CompileState: magma.app.compile, 
	Context: magma.app.compile, 
	ConstructionCaller: magma.app.compile.define, 
	ConstructorHeader: magma.app.compile.define, 
	Definition: magma.app.compile.define, 
	MethodHeader: magma.app.compile.define, 
	Parameter: magma.app.compile.define, 
	Dependency: magma.app.compile, 
	DivideState: magma.app.compile, 
	ImmutableCompileState: magma.app.compile, 
	ImmutableContext: magma.app.compile, 
	ImmutableDivideState: magma.app.compile, 
	ImmutableRegistry: magma.app.compile, 
	ImmutableStack: magma.app.compile, 
	Import: magma.app.compile, 
	Registry: magma.app.compile, 
	Stack: magma.app.compile, 
	Whitespace: magma.app.compile.text, 
	FunctionType: magma.app.compile.type, 
	PrimitiveType: magma.app.compile.type, 
	TemplateType: magma.app.compile.type, 
	Type: magma.app.compile.type, 
	VariadicType: magma.app.compile.type, 
	AccessValue: magma.app.compile.value, 
	Argument: magma.app.compile.value, 
	Caller: magma.app.compile.value, 
	Invokable: magma.app.compile.value, 
	Lambda: magma.app.compile.value, 
	Not: magma.app.compile.value, 
	Operation: magma.app.compile.value, 
	Placeholder: magma.app.compile.value, 
	StringValue: magma.app.compile.value, 
	Symbol: magma.app.compile.value, 
	Value: magma.app.compile.value, 
	CompilerUtils: magma.app, 
	DefiningCompiler: magma.app, 
	DefinitionCompiler: magma.app, 
	FieldCompiler: magma.app, 
	FunctionSegmentCompiler: magma.app, 
	PathSource: magma.app.io, 
	Source: magma.app.io, 
	Location: magma.app, 
	Main: magma.app, 
	PathSources: magma.app, 
	PathTargets: magma.app, 
	Platform: magma.app, 
	RootCompiler: magma.app, 
	Sources: magma.app, 
	Targets: magma.app, 
	ValueCompiler: magma.app
]*/
import { CompileState } from "../../magma/app/compile/CompileState";
import { Tuple2Impl } from "../../magma/api/Tuple2Impl";
import { Value } from "../../magma/app/compile/value/Value";
import { Tuple2 } from "../../magma/api/Tuple2";
import { RootCompiler } from "../../magma/app/RootCompiler";
import { Option } from "../../magma/api/option/Option";
import { Strings } from "../../magma/api/text/Strings";
import { ConstructionCaller } from "../../magma/app/compile/define/ConstructionCaller";
import { Some } from "../../magma/api/option/Some";
import { StringValue } from "../../magma/app/compile/value/StringValue";
import { Not } from "../../magma/app/compile/value/Not";
import { Parameter } from "../../magma/app/compile/define/Parameter";
import { List } from "../../magma/api/collect/list/List";
import { DefiningCompiler } from "../../magma/app/DefiningCompiler";
import { Definition } from "../../magma/app/compile/define/Definition";
import { Iterable } from "../../magma/api/collect/list/Iterable";
import { Lambda } from "../../magma/app/compile/value/Lambda";
import { None } from "../../magma/api/option/None";
import { AccessValue } from "../../magma/app/compile/value/AccessValue";
import { Operation } from "../../magma/app/compile/value/Operation";
export class ValueCompiler {
	static generateValue(tuple: Tuple2<CompileState, Value>): Tuple2Impl<CompileState, string> {
		let state = tuple.left()/*unknown*/;
		let right = tuple.right()/*unknown*/;
		let generated = right.generate()/*unknown*/;
		let s = RootCompiler.generatePlaceholder(RootCompiler.resolve0(state, right).generate())/*unknown*/;
		return new Tuple2Impl<CompileState, string>(state, generated + s)/*unknown*/;
	}
	static parseInvokable(state: CompileState, input: string): Option<Tuple2<CompileState, Value>> {
		return RootCompiler.compileSuffix(Strings.strip(input), ")", (withoutEnd: string) => RootCompiler.compileSplit(RootCompiler.splitFoldedLast(withoutEnd, "", RootCompiler.foldInvocationStarts), (callerWithArgStart: string, args: string) => RootCompiler.compileSuffix(callerWithArgStart, "(", (callerString: string) => RootCompiler.compilePrefix(Strings.strip(callerString), "new ", (type: string) => RootCompiler.compileType(state, type).flatMap((callerTuple: Tuple2<CompileState, string>) => {
			let callerState = callerTuple.right()/*unknown*/;
			let caller = callerTuple.left()/*unknown*/;
			return RootCompiler.assembleInvokable(caller, new ConstructionCaller(callerState), args)/*unknown*/;
		})/*unknown*/).or(() => RootCompiler.parseValue(state, callerString).flatMap((callerTuple: Tuple2<CompileState, Value>) => RootCompiler.assembleInvokable(callerTuple.left(), callerTuple.right(), args)/*unknown*/)/*unknown*/)/*unknown*/)/*unknown*/)/*unknown*/)/*unknown*/;
	}
	static createTextRule(slice: string): (arg0 : CompileState, arg1 : string) => Option<Tuple2<CompileState, Value>> {
		return (state1: CompileState, input1: string) => {
			let stripped = Strings.strip(input1)/*unknown*/;
			return RootCompiler.compilePrefix(stripped, slice, (s: string) => RootCompiler.compileSuffix(s, slice, (s1: string) => new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(state1, new StringValue(s1)))/*unknown*/)/*unknown*/)/*unknown*/;
		}/*unknown*/;
	}
	static parseNot(state: CompileState, input: string): Option<Tuple2<CompileState, Value>> {
		return RootCompiler.compilePrefix(Strings.strip(input), "!", (withoutPrefix: string) => {
			let childTuple = RootCompiler.compileValueOrPlaceholder(state, withoutPrefix)/*unknown*/;
			let childState = childTuple.left()/*unknown*/;
			let child = "!" + childTuple.right()/*unknown*/;
			return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(childState, new Not(child)))/*unknown*/;
		})/*unknown*/;
	}
	static parseLambda(state: CompileState, input: string): Option<Tuple2<CompileState, Value>> {
		return RootCompiler.compileFirst(input, "->", (beforeArrow: string, afterArrow: string) => {
			let strippedBeforeArrow = Strings.strip(beforeArrow)/*unknown*/;
			return RootCompiler.compilePrefix(strippedBeforeArrow, "(", (withoutStart: string) => {
				return RootCompiler.compileSuffix(withoutStart, ")", (withoutEnd: string) => {
					return RootCompiler.parseValues(state, withoutEnd, (state1: CompileState, s: string) => {
						return RootCompiler.parseParameter(state1, s)/*unknown*/;
					}).flatMap((paramNames: Tuple2<CompileState, List<Parameter>>) => {
						return compileLambdaWithParameterNames(paramNames.left(), DefiningCompiler.retainDefinitionsFromParameters(paramNames.right()), afterArrow)/*unknown*/;
					})/*unknown*/;
				})/*unknown*/;
			})/*unknown*/;
		})/*unknown*/;
	}
	static compileLambdaWithParameterNames(state: CompileState, paramNames: Iterable<Definition>, afterArrow: string): Option<Tuple2<CompileState, Value>> {
		let strippedAfterArrow = Strings.strip(afterArrow)/*unknown*/;
		return RootCompiler.compilePrefix(strippedAfterArrow, "{", (withoutContentStart: string) => RootCompiler.compileSuffix(withoutContentStart, "}", (withoutContentEnd: string) => {
			let compileState: CompileState = state.enterDepth()/*unknown*/;
			let statementsTuple = RootCompiler.compileFunctionStatements(compileState.mapStack(stack1 -  > stack1.defineAll(paramNames)), withoutContentEnd)/*unknown*/;
			let statementsState = statementsTuple.left()/*unknown*/;
			let statements = statementsTuple.right()/*unknown*/;
			let exited = statementsState.exitDepth()/*unknown*/;
			return assembleLambda(exited, paramNames, "{" + statements + exited.createIndent() + "}")/*unknown*/;
		})/*unknown*/).or(() => RootCompiler.compileValue(state, strippedAfterArrow).flatMap((tuple: Tuple2<CompileState, string>) => assembleLambda(tuple.left(), paramNames, tuple.right())/*unknown*/)/*unknown*/)/*unknown*/;
	}
	static assembleLambda(exited: CompileState, paramNames: Iterable<Definition>, content: string): Option<Tuple2<CompileState, Value>> {
		return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(exited, new Lambda(paramNames, content)))/*unknown*/;
	}
	static createOperatorRule(infix: string): (arg0 : CompileState, arg1 : string) => Option<Tuple2<CompileState, Value>> {
		return createOperatorRuleWithDifferentInfix(infix, infix)/*unknown*/;
	}
	static createAccessRule(infix: string): (arg0 : CompileState, arg1 : string) => Option<Tuple2<CompileState, Value>> {
		return (state: CompileState, input: string) => RootCompiler.compileLast(input, infix, (childString: string, rawProperty: string) => {
			let property = Strings.strip(rawProperty)/*unknown*/;
			if (!RootCompiler/*unknown*/.isSymbol(property)/*unknown*/){
				return new None<Tuple2<CompileState, Value>>()/*unknown*/;
			}
			return RootCompiler.parseValue(state, childString).flatMap((childTuple: Tuple2<CompileState, Value>) => {
				let childState = childTuple.left()/*unknown*/;
				let child = childTuple.right()/*unknown*/;
				return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(childState, new AccessValue(child, property)))/*unknown*/;
			})/*unknown*/;
		})/*unknown*//*unknown*/;
	}
	static createOperatorRuleWithDifferentInfix(sourceInfix: string, targetInfix: string): (arg0 : CompileState, arg1 : string) => Option<Tuple2<CompileState, Value>> {
		return (state1: CompileState, input1: string) => RootCompiler.compileSplit(RootCompiler.splitFolded(input1, RootCompiler.foldOperator(sourceInfix), (divisions: List<string>) => RootCompiler.selectFirst(divisions, sourceInfix)/*unknown*/), (leftString: string, rightString: string) => RootCompiler.parseValue(state1, leftString).flatMap((leftTuple: Tuple2<CompileState, Value>) => RootCompiler.parseValue(leftTuple.left(), rightString).flatMap((rightTuple: Tuple2<CompileState, Value>) => {
			let left = leftTuple.right()/*unknown*/;
			let right = rightTuple.right()/*unknown*/;
			return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(rightTuple.left(), new Operation(left, targetInfix, right)))/*unknown*/;
		})/*unknown*/)/*unknown*/)/*unknown*//*unknown*/;
	}
}

File: src/java/magma/app/ValueCompiler.java
package magma.app;

import jvm.api.collect.list.Lists;
import magma.api.Tuple2;
import magma.api.Tuple2Impl;
import magma.api.collect.Iters;
import magma.api.collect.head.HeadedIter;
import magma.api.collect.head.RangeHead;
import magma.api.collect.list.Iterable;
import magma.api.collect.list.List;
import magma.api.collect.list.ListCollector;
import magma.api.option.None;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.api.text.Characters;
import magma.api.text.Strings;
import magma.app.compile.CompileState;
import magma.app.compile.DivideState;
import magma.app.compile.Stack;
import magma.app.compile.compose.Composable;
import magma.app.compile.compose.PrefixComposable;
import magma.app.compile.compose.SplitComposable;
import magma.app.compile.compose.SuffixComposable;
import magma.app.compile.define.ConstructionCaller;
import magma.app.compile.define.Definition;
import magma.app.compile.define.Parameter;
import magma.app.compile.fold.ValueFolder;
import magma.app.compile.rule.OrRule;
import magma.app.compile.rule.Rule;
import magma.app.compile.select.FirstSelector;
import magma.app.compile.select.Selector;
import magma.app.compile.split.FoldingSplitter;
import magma.app.compile.split.Splitter;
import magma.app.compile.type.Type;
import magma.app.compile.value.AccessValue;
import magma.app.compile.value.Argument;
import magma.app.compile.value.Caller;
import magma.app.compile.value.Invokable;
import magma.app.compile.value.Lambda;
import magma.app.compile.value.Not;
import magma.app.compile.value.Operation;
import magma.app.compile.value.Placeholder;
import magma.app.compile.value.StringValue;
import magma.app.compile.value.Symbol;
import magma.app.compile.value.Value;
import magma.app.compile.fold.OperatorFolder;
import magma.app.compile.locate.FirstLocator;
import magma.app.compile.select.LastSelector;
import magma.app.compile.split.LocatingSplitter;

import java.util.function.Function;

public final class ValueCompiler {
    static Tuple2Impl<CompileState, String> generateValue(Tuple2<CompileState, Value> tuple) {
        var state = tuple.left();
        var right = tuple.right();
        var generated = right.generate();
        var s = Placeholder.generatePlaceholder(ValueCompiler.resolve(state, right).generate());
        return new Tuple2Impl<CompileState, String>(state, generated + s);
    }

    static Option<Tuple2<CompileState, Value>> parseInvokable(CompileState state, String input) {
        return new SuffixComposable<Tuple2<CompileState, Value>>(")", (String withoutEnd) -> {
            return new SplitComposable<Tuple2<CompileState, Value>>((String withoutEnd0) -> {
                    Selector selector = new LastSelector("");
                    return new FoldingSplitter((DivideState state1, char c) -> {
                        return ValueCompiler.foldInvocationStarts(state1, c);
                    }, selector).apply(withoutEnd0);
                }, Composable.toComposable((String callerWithArgStart, String args) -> {
                    return new SuffixComposable<Tuple2<CompileState, Value>>("(", (String callerString) -> {
                        return new PrefixComposable<Tuple2<CompileState, Value>>("new ", (String type) -> {
                            return TypeCompiler.compileType(state, type).flatMap((Tuple2<CompileState, String> callerTuple1) -> {
                                var callerState = callerTuple1.right();
                                var caller = callerTuple1.left();
                                return ValueCompiler.assembleInvokable(caller, new ConstructionCaller(callerState), args);
                            });
                        }).apply(Strings.strip(callerString)).or(() -> {
                            return parseValue(state, callerString).flatMap((Tuple2<CompileState, Value> callerTuple) -> {
                                return ValueCompiler.assembleInvokable(callerTuple.left(), callerTuple.right(), args);
                            });
                        });
                    }).apply(callerWithArgStart);
                })).apply(withoutEnd);
        }).apply(Strings.strip(input));
    }

    static Rule<Value> createTextRule(String slice) {
        return (CompileState state1, String input1) -> {
            var stripped = Strings.strip(input1);
            return new PrefixComposable<Tuple2<CompileState, Value>>(slice, (String s) -> {
                return new SuffixComposable<Tuple2<CompileState, Value>>(slice, (String s1) -> {
                                return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(state1, new StringValue(s1)));
                            }).apply(s);
            }).apply(stripped);
        };
    }

    static Option<Tuple2<CompileState, Value>> parseNot(CompileState state, String input) {
        return new PrefixComposable<Tuple2<CompileState, Value>>("!", (String withoutPrefix) -> {
            var childTuple = ValueCompiler.compileValueOrPlaceholder(state, withoutPrefix);
            var childState = childTuple.left();
            var child = "!" + childTuple.right();
            return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(childState, new Not(child)));
        }).apply(Strings.strip(input));
    }

    static Option<Tuple2<CompileState, Value>> parseLambda(CompileState state, String input) {
        Splitter splitter = new LocatingSplitter("->", new FirstLocator());
        return new SplitComposable<Tuple2<CompileState, Value>>(splitter, Composable.toComposable((String beforeArrow, String afterArrow) -> {
            var strippedBeforeArrow = Strings.strip(beforeArrow);
            return new PrefixComposable<Tuple2<CompileState, Value>>("(", (String withoutStart) -> {
                return new SuffixComposable<Tuple2<CompileState, Value>>(")", (String withoutEnd) -> {
                    return values((CompileState state1, String s) -> {
                        return DefiningCompiler.parseParameter(state1, s);
                    }).apply(state, withoutEnd).flatMap((Tuple2<CompileState, List<Parameter>> paramNames) -> {
                                    return ValueCompiler.compileLambdaWithParameterNames(paramNames.left(), DefiningCompiler.retainDefinitionsFromParameters(paramNames.right()), afterArrow);
                                });
                            }).apply(withoutStart);
            }).apply(strippedBeforeArrow);
        })).apply(input);
    }

    private static Option<Tuple2<CompileState, Value>> compileLambdaWithParameterNames(CompileState state, Iterable<Definition> paramNames, String afterArrow) {
        var strippedAfterArrow = Strings.strip(afterArrow);
        return new PrefixComposable<Tuple2<CompileState, Value>>("{", (String withoutContentStart) -> {
            return new SuffixComposable<Tuple2<CompileState, Value>>("}", (String withoutContentEnd) -> {
                    CompileState compileState = state.enterDepth();
                    var statementsTuple = FunctionSegmentCompiler.compileFunctionStatements(compileState.mapStack((Stack stack1) -> {
                        return stack1.defineAll(paramNames);
                    }), withoutContentEnd);
                    var statementsState = statementsTuple.left();
                    var statements = statementsTuple.right();

                    var exited = statementsState.exitDepth();
                    return ValueCompiler.assembleLambda(exited, paramNames, "{" + statements + exited.createIndent() + "}");
                }).apply(withoutContentStart);
        }).apply(strippedAfterArrow).or(() -> {
            return ValueCompiler.compileValue(state, strippedAfterArrow).flatMap((Tuple2<CompileState, String> tuple) -> {
                return ValueCompiler.assembleLambda(tuple.left(), paramNames, tuple.right());
            });
        });
    }

    private static Option<Tuple2<CompileState, Value>> assembleLambda(CompileState exited, Iterable<Definition> paramNames, String content) {
        return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(exited, new Lambda(paramNames, content)));
    }

    static Rule<Value> createOperatorRule(String infix) {
        return ValueCompiler.createOperatorRuleWithDifferentInfix(infix, infix);
    }

    static Rule<Value> createAccessRule(String infix) {
        return (CompileState state, String input) -> {
            return SplitComposable.compileLast(input, infix, (String childString, String rawProperty) -> {
                var property = Strings.strip(rawProperty);
                if (!ValueCompiler.isSymbol(property)) {
                    return new None<Tuple2<CompileState, Value>>();
                }

                return parseValue(state, childString).flatMap((Tuple2<CompileState, Value> childTuple) -> {
                    var childState = childTuple.left();
                    var child = childTuple.right();
                    return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(childState, new AccessValue(child, property)));
                });
            });
        };
    }

    static Rule<Value> createOperatorRuleWithDifferentInfix(String sourceInfix, String targetInfix) {
        return (CompileState state1, String input1) -> {
            return new SplitComposable<Tuple2<CompileState, Value>>((String slice) -> {
                    return new FoldingSplitter(new OperatorFolder(sourceInfix), (List<String> divisions) -> {
                            return new FirstSelector(sourceInfix).select(divisions);
                        }).apply(slice);
                }, Composable.toComposable((String leftString, String rightString) -> {
                    return parseValue(state1, leftString).flatMap((Tuple2<CompileState, Value> leftTuple) -> {
                        return parseValue(leftTuple.left(), rightString).flatMap((Tuple2<CompileState, Value> rightTuple) -> {
                            var left = leftTuple.right();
                            var right = rightTuple.right();
                            return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(rightTuple.left(), new Operation(left, targetInfix, right)));
                        });
                    });
                })).apply(input1);
        };
    }

    static Option<Tuple2<CompileState, Value>> parseSymbol(CompileState state, String input) {
        var stripped = Strings.strip(input);
        if (ValueCompiler.isSymbol(stripped)) {
            var withImport = TypeCompiler.addResolvedImportFromCache0(state, stripped);
            return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(withImport, new Symbol(stripped)));
        }
        else {
            return new None<Tuple2<CompileState, Value>>();
        }
    }

    static boolean isSymbol(String input) {
        var query = new HeadedIter<Integer>(new RangeHead(Strings.length(input)));
        return query.allMatch((Integer index) -> {
            return ValueCompiler.isSymbolChar(index, input.charAt(index));
        });
    }

    private static boolean isSymbolChar(int index, char c) {
        return '_' == c
                || Characters.isLetter(c)
                || (0 != index && Characters.isDigit(c));
    }

    private static boolean isNumber(String input) {
        var query = new HeadedIter<Integer>(new RangeHead(Strings.length(input)));
        return query.map(input::charAt).allMatch((Character c) -> {
            return Characters.isDigit(c);
        });
    }

    private static Type resolve(CompileState state, Value value) {
        return switch (value) {
            case AccessValue accessValue -> accessValue.resolve(state);
            case Invokable invokable -> invokable.resolve(state);
            case Lambda lambda -> lambda.resolve(state);
            case Not not -> not.resolve(state);
            case Operation operation -> operation.resolve(state);
            case Placeholder placeholder -> placeholder.resolve(state);
            case StringValue stringValue -> stringValue.resolve(state);
            case Symbol symbol -> symbol.resolve(state);
        };
    }

    static Option<Tuple2<CompileState, Value>> parseNumber(CompileState state, String input) {
        var stripped = Strings.strip(input);
        if (ValueCompiler.isNumber(stripped)) {
            return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(state, new Symbol(stripped)));
        }
        else {
            return new None<Tuple2<CompileState, Value>>();
        }
    }

    static Tuple2<CompileState, String> compileValueOrPlaceholder(CompileState state, String input) {
        return ValueCompiler.compileValue(state, input).orElseGet(() -> {
            return new Tuple2Impl<CompileState, String>(state, Placeholder.generatePlaceholder(input));
        });
    }

    static Option<Tuple2<CompileState, String>> compileValue(CompileState state, String input) {
        return parseValue(state, input).map((Tuple2<CompileState, Value> tuple) -> {
            return ValueCompiler.generateValue(tuple);
        });
    }

    private static Option<Tuple2<CompileState, Argument>> parseArgument(CompileState state1, String input) {
        return parseValue(state1, input)
                .map((Tuple2<CompileState, Value> tuple) -> {
                    return new Tuple2Impl<CompileState, Argument>(tuple.left(), tuple.right());
                });
    }

    private static Caller transformCaller(CompileState state, Caller oldCaller) {
        return oldCaller.findChild().flatMap((Value parent) -> {
            var parentType = ValueCompiler.resolve(state, parent);
            if (parentType.isFunctional()) {
                return new Some<Caller>(parent);
            }

            return new None<Caller>();
        }).orElse(oldCaller);
    }

    private static DivideState foldInvocationStarts(DivideState state, char c) {
        var appended = state.append(c);
        if ('(' == c) {
            var entered = appended.enter();
            if (entered.isShallow()) {
                return entered.advance();
            }
            else {
                return entered;
            }
        }

        if (')' == c) {
            return appended.exit();
        }

        return appended;
    }

    private static Option<Tuple2<CompileState, Value>> assembleInvokable(CompileState state, Caller oldCaller, String argsString) {
        return values((CompileState state1, String s) -> {
            return ValueCompiler.parseArgument(state1, s);
        }).apply(state, argsString).flatMap((Tuple2<CompileState, List<Argument>> argsTuple) -> {
            var argsState = argsTuple.left();
            var args = retain(argsTuple.right(), (Argument argument) -> {
                return argument.toValue();
            });

            var newCaller = ValueCompiler.transformCaller(argsState, oldCaller);
            return new Some<Tuple2<CompileState, Value>>(new Tuple2Impl<CompileState, Value>(argsState, new Invokable(newCaller, args)));
        });
    }

    static <T, R> Iterable<R> retain(Iterable<T> args, Function<T, Option<R>> mapper) {
        return args.iter()
                .map(mapper)
                .flatMap(Iters::fromOption)
                .collect(new ListCollector<R>());
    }

    public static Option<Tuple2<CompileState, Value>> parseValue(CompileState state, String input) {
        return new OrRule<Value>(Lists.of(
                ValueCompiler::parseLambda,
                createOperatorRule("+"),
                createOperatorRule("-"),
                createOperatorRule("<="),
                createOperatorRule("<"),
                createOperatorRule("&&"),
                createOperatorRule("||"),
                createOperatorRule(">"),
                createOperatorRule(">="),
                ValueCompiler::parseInvokable,
                createAccessRule("."),
                createAccessRule("::"),
                ValueCompiler::parseSymbol,
                ValueCompiler::parseNot,
                ValueCompiler::parseNumber,
                createOperatorRuleWithDifferentInfix("==", "==="),
                createOperatorRuleWithDifferentInfix("!=", "!=="),
                createTextRule("\""),
                createTextRule("'")
        )).apply(state, input);
    }

    public static <T> Rule<List<T>> values(Rule<T> mapper) {
        return new DivideRule<>(new ValueFolder(), mapper);
    }
}